hiku/federation/endpoint.py

```diff
@@ -0,0 +1,155 @@
+from abc import abstractmethod
+from asyncio import gather
+from contextlib import contextmanager
+from typing import (
+    List,
+    Dict,
+    Any,
+)
+
+from .utils import get_keys
+from .introspection import (
+    FederatedGraphQLIntrospection,
+    AsyncFederatedGraphQLIntrospection,
+    is_introspection_query,
+    extend_with_federation,
+)
+from .validate import validate
+
+from hiku.denormalize.graphql import (
+    DenormalizeGraphQL,
+)
+from hiku.endpoint.graphql import (
+    BaseGraphQLEndpoint,
+    _type_names,
+    _switch_graph,
+    GraphQLError,
+    _StripQuery,
+)
+from hiku.graph import Graph
+from hiku.query import Node
+from hiku.result import Proxy, Reference
+
+
+def _process_query(graph, query):
+    stripped_query = _StripQuery().visit(query)
+    errors = validate(graph, stripped_query)
+    if errors:
+        raise GraphQLError(errors=errors)
+    else:
+        return stripped_query
+
+
+def denormalize_entities(
+    graph: Graph,
+    query: Node,
+    result: Proxy,
+    type_name: str
+) -> List[Dict[str, Any]]:
+
+    entities_link = query.fields_map['_entities']
+    node = entities_link.node
+    representations = entities_link.options['representations']
+
+    entities = []
+    for r in representations:
+        typename = r['__typename']
+        for key in get_keys(graph, typename):
+            if key not in r:
+                continue
+            ident = r[key]
+            result.__ref__ = Reference(typename, ident)
+            data = DenormalizeGraphQL(graph, result, type_name).process(node)
```

Comment:
Whoops. This is broken (works by coincedence). Turns out, I had a link `cart` in Root node, and a query `cart { id }`, so `self._type[-1].__field_types__[obj.name]` worked fine. This should be fixed somehow.

Reply:
fixed by using type from representations as root type for denormalization instead of `__root__`